Downloading folder settings

samytichadou opened the pull request:
This PR adds some control over the download locations of the previz sounds and strip source sounds :
- In the addon preferences, 2 new properties, one is a folder pattern (name) if the user choose to store some datas next to the blend file, the other is a common folder where downloads are stored if user choose common folder option
![image](https://user-images.githubusercontent.com/26276003/146396551-b198e606-c5ab-4b5f-9051-639e01dc6902.png)
- In the UI, two options, one for the previz location, the other for the download location. Both are 2 entries enum prop with "Common" and "Project" options to store sound files in a folder alongside project, or common one (both specified in addon prefs)
![image](https://user-images.githubusercontent.com/26276003/146396876-f0942434-fac5-4037-b8d2-b4024d347212.png)
- I also fixed the "already existing file" behavior in play and add sound operators by using name of sound file from the addon_data list of sounds, so the addon does not re download the sound if it finds it.
- The addon shows warning message if the sound file is supposed to be saved alongside blend file, but blend file is not saved
- The addon shows warning message if the sound file is supposed to be saved alongside blend file, but folder name pattern from prefs is blank
- The addon creates recursively the folders on-the-fly when needed 
- Default for the common folder is located in blender user datafiles folder
- I bumped the version and adds tin2tin and myself as authors in the bl_info, hope this is ok
- Added a few prints in the play and add operators (downloading, playing, adding strips)

I tried to keep the code consistent without altering existing code, i added 2 commonly used functions at the start of the freesound.py file and a few import `from os.path`

Cheers !

Commit:
init folder download selection

use folders, mkdir if non existent

Commit:
add scene property for the download behavior

Commit:
get prefs function

Commit:
filepath building init

Commit:
create folder function

Commit:
build download filepath in function

Commit:
correct os import and use

Commit:
fix already downloaded download in add operator

use sound name from addon_data.freesound_list tp check existing file

Commit:
add preview_location prop and use it in preview operator

Commit:
get possible errors in file download locations

Commit:
tidy addon prefs

Commit:
naming

Commit:
syntax and removing unused prefs function

samytichadou commented:
I am using currently the addon, and i'm not sure if the best solution is to have the locations properties (download and previz) in the ui and scene settings, or in the addon preferences ? My initial take was it could be useful to change it regularly depending on the project, but not so sure now, what do you think ?

iwkse commented:
Hi, thanks for this PR!
I think it would be better to add it in the configuration panel, but I would rename the label for the paths. It's not clear one is for the preview. I would explicitly write preview and download.

iwkse commented:
Thinking more about it, adding it directly to the freesound pane could be more flexible and could be possible to have separate folders for each blend file. Could be nice I think.

tin2tin commented:
If the folder settings should be exposed in the sidebar, maybe it would be better to place them in a properties section after the search and the results sections(eg. in the end)? (My initial thought was also to place them in the configuration panel)

(Btw. I got stuck on the #18 exposing of waveform previews received from freesound. I guess it should be possible to expose them in the list control. Do any of you guys know how to do that? In #18 I got all of the loading etc. of the waveform previews)

iwkse commented:
> If the folder settings should be exposed in the sidebar, maybe it would be better to place them in a properties section after the search and the results sections(eg. in the end)? (My initial thought was also to place them in the configuration panel)
> 
> (Btw. I got stuck on the #18 exposing of waveform previews received from freesound. I guess it should be possible to expose them in the list control. Do any of you guys know how to do that? In #18 I got all of the loading etc. of the waveform previews)

Seems like a good idea. Sorry for the waveform but I really had little to no time lately. Hopefully next days it will be better and I can look into fixing the pending issues

samytichadou commented:
> 

So I guess i'll add these two new properties under a "scene settings" subpanel ? and maybe create a "search" subpanel also to have all freesound related ui clean and organized ? Would it be acceptable solution for both of you ?
But to be clear, it would not be possible to have multiple folders per blend file, if blend files are in the same locations, "alongside" folder will always be blend_file_location/freesound_folder , with freesound_folder being the pattern set in the preferences. Are we all ok with this behavior ? Seems efficient and logic enough to me
Cheers

iwkse commented:
>     * I bumped the version and adds tin2tin and myself as authors in the bl_info, hope this is ok
As for this, I'd leave myself as author, just for simplicity. Otherwise such field will have to be filled with all possible developer opening a PR and contributing to the addon.
I'm really thankful for the contributions and I'm glad to evaluate and accept. I'm quite open instead to add in the README a section for contributions.

samytichadou commented:
> > ```
> > * I bumped the version and adds tin2tin and myself as authors in the bl_info, hope this is ok
> > ```
> 
> As for this, I'd leave myself as author, just for simplicity. Otherwise such field will have to be filled with all possible developer opening a PR and contributing to the addon. I'm really thankful for the contributions and I'm glad to evaluate and accept. I'm quite open instead to add in the README a section for contributions.

ok, i always did that and find it kind of logic, but do as you want

Commit:
remove unused import

Commit:
poll out search subpanel if no credentials and ui improvement

Commit:
proper sidebar category

Commit:
settings subpanel opened by default

samytichadou commented:
Just added the subpanels : 
![image](https://user-images.githubusercontent.com/26276003/146695719-5c83fd57-710d-48de-956b-cd32ac9c8a25.png)
The download location part is in a box, to be able to easily add other scene settings if needed.
The search subpanel is polled out if credentials are missing

I also added some informations when freesound credentials are not correct : 
![image](https://user-images.githubusercontent.com/26276003/146695734-c5f74697-8fd7-4a01-9588-5ce7c58fda7d.png)

I added a common class for freesound panels, and removed a few unused import.

The "conflict" detected by github is about this common class, but you can test, np about it, just another way of doing it.

iwkse commented:
No, the conflict it's because you didn't pull changes before committing. You can rebase and force push and it will be fixed

Commit:
conflict resolution

samytichadou commented:
> No, the conflict it's because you didn't pull changes before committing. You can rebase and force push and it will be fixed

Oops, didn't check, just fixed it

iwkse commented on __init__.py:
```diff
@@ -48,12 +68,18 @@ def draw(self, context):
     "name": "Freesound",
-    "author": "Salvatore De Paolis",
-    "version": (2, 0),
+    "author": "Salvatore De Paolis, tin2tin, tonton",
```
Please revert this

samytichadou replied:
reverted the author, keeped the version bumped, 2 -> 2.1 ok with that ?

iwkse commented on freesound.py:
```diff
@@ -88,6 +88,7 @@ def execute(self, context):
                     soundfile = sound_info.retrieve_preview(dirname(realpath(__file__)),
                                                     sound_info.name,
                                                     addon_data.high_quality)
+                # soundfile = sound filepath
```
Remove this

iwkse commented on freesound.py:
```diff
@@ -343,6 +344,7 @@ def execute(self, context):
                                             sound_info.name,
                                             addon_data.high_quality)
         addon_data.soundfile = soundfile
+        # soundfile = sound filepath
```
Remove this

iwkse commented on freesound.py:
```diff
@@ -203,6 +203,16 @@ def update_max(self, context):
 
+    download_location: EnumProperty(
+        items = [
+            ('PROJECT', 'Alongside Project', 'Alongside Project'),
```
I'd use Within Project label

samytichadou replied:
"Within project" seems less clear to me, the sounds are not in the blend, they are located "alongside" or "near", maybe "project directory" would also be more explicit, don't you think?

iwkse replied:
Right, I think project directory sounds better

iwkse commented on freesound.py:
```diff
@@ -10,6 +10,12 @@
 import aud
 from . import freesound_api
 
+# get addon preferences
```
Remove this, it's already clear from the method's name

iwkse commented on freesound.py:
```diff
@@ -333,7 +339,10 @@ class Freesound_Add(btypes.Operator):
     bl_options = {'REGISTER', 'UNDO'}
 
+    # poll out if project location and not saved
+
```
Remove whiteline

iwkse commented on freesound.py:
```diff
@@ -14,13 +14,27 @@
         os.makedirs(folderpath, exist_ok=True)
     return folderpath
 
+# build download filepath
```
Remove this, already clear from the method's name

iwkse commented on __init__.py:
```diff
@@ -46,8 +46,8 @@ class ApiAddonPreferences(AddonPreferences):
+        name = "Common folder path for download",
+        description = "Common folder path where freesound downloads are stored",
         subtype = 'DIR_PATH',
         default = os.path.join(bpy.utils.user_resource('DATAFILES'), "freesound_downloads")
```
Just use join

samytichadou replied:
done, for the record, I think it could be better practice to import the entire library for better readability, specially for open source code (a contributor does not have to know about os module or check in a code editor to get what is "join", "os.path.join" is more explicit (and the method is used only once here, no real gain in characters)
just my point of view about it.

iwkse commented on freesound.py:
```diff
@@ -38,6 +34,7 @@ def build_download_filepath(download_location, prefs, file_name):
     
     return sound_filepath
 
+
```
Remove newline

iwkse commented on ui.py:
```diff
@@ -18,9 +18,7 @@
-from . import freesound_api
-from . import freesound
-import datetime
+
```
Remove newline

iwkse commented on ui.py:
```diff
@@ -23,12 +23,13 @@
     """Creates a Panel in the Object properties window"""
     bl_label = "Freesound"
-
+    
```
Remove spaces

samytichadou replied:
36139ac3668b23903c38d5f6c093736daca20c52

iwkse reviewed (commented):
Please remove the merge

samytichadou commented:
> Please remove the merge

Not sure where the merge comes from, rebase seems to did this, I ll see what I can do

Commit:
syntax

Commit:
syntax

Commit:
remove authors

Commit:
specific import from os

Commit:
change naming of download locations props

samytichadou commented:
all done

iwkse commented:
> all done

Thanks very much, I'm going to merge it
